blockchain/chain.go

```diff
@@ -703,6 +708,9 @@ func (b *BlockChain) connectBlock(node *blockNode, block, parent *dcrutil.Block,
 	b.stateSnapshot = state
 	b.stateLock.Unlock()
 
+	// Notify the spend pruner of the connected block.
+	go b.spendPruner.NotifyConnectedBlock(block.Hash())
```

Comment:
Agreed with @davecgh's assessment here that this could result in a spend journal entry being removed for a block on the main chain, which would subsequently result in a [panic](https://github.com/decred/dcrd/blob/598bf66f24a671f90c0c1b3e190103159e2cf4ab/blockchain/chainio.go#L764) if that block needs to be disconnected.

Reply:
To follow up, I was able to cause a failure randomly by modifying the code to intentionally add delays in just the right way so both events happen around the same time and just keep doing it until it eventually (because channel selection is random) hits the issue I described.  I suspect it could be reproduced a little more easily by using a separate mutex and synchronizing things in the "wrong" order such that it's constantly contending on the two channels, but I didn't spend any more time since I was able to reproduce and prove it is an issue.

I believe a way to fix things properly is to make two changes:

1) Combine `pruneCh` and `connectCh` into a single channel with a discriminant/type to specify which action is to be taken in order to ensure all modifications are properly ordered
2) Move the actual removal logic to a method on `blockchain` (that is provided to the pruner as a callback) such that said method takes the `processLock` which ensures things can't change out from under it.  Then, inside that function, only remove the spend entry when it is **not** part of the main chain and ignore it otherwise.

The combination of those two things will ensure a consistent state despite the indexing and pruning being done independently and asynchronously because former will ensure the pruner code itself will remove entries for the cases it knows about and the latter will do the same for the cases it does not.

Reply:
Noted, thanks for the review.